File: BI-TJV/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/food/FoodAdapter.java
package cz.wrzecond.restsys.food;

import android.content.Intent;
import android.graphics.Color;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.AppCompatImageButton;

import java.util.List;
import java.util.Locale;

import cz.wrzecond.restsys.R;
import cz.wrzecond.restsys.task.HttpMethod;
import cz.wrzecond.restsys.task.HttpRequest;
import cz.wrzecond.restsys.Adapter;
import cz.wrzecond.restsys.allergen.AllergenActivity;
import wrzecond.food.TjvFoodReadDTO;

public class FoodAdapter extends Adapter<TjvFoodReadDTO> {

    // Properties
    private FoodActivity activity;
    private final boolean selection;

    // Constructors
    public FoodAdapter (FoodActivity activity, @NonNull List<TjvFoodReadDTO> dtos, boolean selection) {
        super (activity, R.layout.item_food, dtos);
        this.activity = activity;
        this.selection = selection;
    }
    public FoodAdapter (FoodActivity activity, @StringRes int errorResId) {
        super (activity, R.layout.item_food, errorResId);
        this.activity = activity;
        this.selection = false;
    }

    @Override
    protected Adapter<TjvFoodReadDTO>.ViewHolder createViewHolder (View view) {
        return new ViewHolder(view);
    }

    protected class ViewHolder extends Adapter<TjvFoodReadDTO>.ViewHolder {

        // === Properties
        private View view;
        private TextView foodName;
        private TextView foodPrice;
        private TextView foodCount;

        private TextView allergenText;
        private AppCompatImageButton allergenInfo;

        private ImageView foodEdit;
        private ImageView foodDelete;

        public ViewHolder(@NonNull View view) { super(view); }

        @Override
        protected void initView (@NonNull View view) {
            this.view = view;
            this.foodName = view.findViewById(R.id.food_name);
            this.foodPrice = view.findViewById(R.id.food_price);
            this.foodCount = view.findViewById(R.id.food_count);

            this.allergenText = view.findViewById(R.id.allergen_text);
            this.allergenInfo = view.findViewById(R.id.allergen_info);

            this.foodEdit = view.findViewById(R.id.food_edit);
            this.foodDelete = view.findViewById(R.id.food_delete);
        }

        @Override
        protected void update ( @NonNull TjvFoodReadDTO dto ) {
            foodName.setText(dto.getName());
            foodPrice.setText(String.format(Locale.getDefault(), "%d,-", dto.getPrice()));

            if (dto.getAllergens().size() > 0) {
                StringBuilder builder = new StringBuilder(activity.getString(R.string.food_allergens) + ":");
                dto.getAllergens().forEach( i -> { builder.append(" "); builder.append(i); } );
                allergenText.setText(builder.toString());
            }
            else {
                allergenText.setVisibility(View.GONE);
                allergenInfo.setVisibility(View.GONE);
            }

            // Launch allergen info
            allergenInfo.setOnClickListener( v -> {
                Intent intent = new Intent(activity, AllergenActivity.class);
                intent.putExtra(AllergenActivity.EXTRA_FOOD_ID, dto.getId());
                activity.startActivity(intent);
            } );

            // Selection - select food on click
            if ( selection ) {
                view.setOnClickListener( v -> {
                    // Selected, add one
                    if ( activity.selectedFood.containsKey(dto.getId()) ) {
                        Integer originalCount = activity.selectedFood.get(dto.getId());
                        if (originalCount == null)
                            return;

                        activity.selectedFood.put(dto.getId(), originalCount + 1);
                        foodCount.setText(String.format(Locale.getDefault(), "%dx", originalCount + 1));
                    }
                    else {
                        activity.selectedFood.put(dto.getId(), 1);
                        foodCount.setVisibility(View.VISIBLE);
                        foodCount.setText(String.format(Locale.getDefault(), "%dx", 1));
                        foodDelete.setVisibility(View.VISIBLE);
                        v.setBackgroundColor(Color.argb(127, 0, 255, 0));
                    }

                    activity.findViewById(R.id.fab).setVisibility(activity.selectedFood.isEmpty() ? View.GONE : View.VISIBLE);
                } );

                foodDelete.setOnClickListener( v -> {
                    // Decrement by one
                    if ( activity.selectedFood.containsKey(dto.getId()) ) {
                        Integer originalCount = activity.selectedFood.get(dto.getId());
                        if (originalCount == null)
                            return;

                        if (originalCount == 1) {
                            activity.selectedFood.remove(dto.getId());
                            view.setBackgroundColor(Color.rgb(255, 255, 255));
                            foodCount.setVisibility(View.GONE);
                            foodDelete.setVisibility(View.INVISIBLE);
                        }
                        else {
                            activity.selectedFood.put(dto.getId(), originalCount - 1);
                            foodCount.setText(String.format(Locale.getDefault(), "%dx", originalCount - 1));
                        }
                    }

                    activity.findViewById(R.id.fab).setVisibility(activity.selectedFood.isEmpty() ? View.GONE : View.VISIBLE);
                } );

                // Already pre-selected
                if ( activity.selectedFood.containsKey(dto.getId()) ) {
                    foodCount.setVisibility(View.VISIBLE);
                    foodCount.setText(String.format(Locale.getDefault(), "%dx", activity.selectedFood.get(dto.getId())));

                    foodEdit.setVisibility(View.GONE);
                    foodDelete.setVisibility(View.VISIBLE);
                    view.setBackgroundColor(Color.argb(127, 0, 255, 0));
                }
                else {
                    foodEdit.setVisibility(View.GONE);
                    foodDelete.setVisibility(View.INVISIBLE);
                }

                activity.findViewById(R.id.fab).setVisibility(activity.selectedFood.isEmpty() ? View.GONE : View.VISIBLE);
                return;
            }

            // Only for admins
            if ( !activity.getService().isAdmin() || activity.getIntent() == null ||
                    !activity.getIntent().hasExtra(FoodActivity.EXTRA_SHOW_NAVIGATION) ) {
                foodEdit.setVisibility(View.GONE);
                foodDelete.setVisibility(View.INVISIBLE);
                return;
            }

            foodEdit.setOnClickListener( v -> {
                AlertDialog dialog = new AlertDialog.Builder(activity)
                    .setView(R.layout.create_food)
                    .create();

                dialog.setOnShowListener(activity.createListener(dialog, dto));
                dialog.show();
            } );

            foodDelete.setOnClickListener( v ->
                new AlertDialog.Builder(activity)
                    .setTitle(R.string.food_delete)
                    .setMessage(activity.getString(R.string.food_delete_confirm, dto.getName()))
                    .setNegativeButton(R.string.button_dismiss, null)
                    .setPositiveButton(R.string.button_delete, ((dialog, which) -> {
                        // Schedule request
                        HttpRequest request = new HttpRequest("/food/" + dto.getId(), HttpMethod.DELETE);
                        activity.getService().scheduleTask( (response) -> {
                            // Show message and close
                            Toast.makeText(activity,
                                response.getStatus().isSuccess() ? activity.getString(R.string.delete_success) :
                                    activity.getString(R.string.delete_failure, response.getStatus().name()),
                            Toast.LENGTH_SHORT).show();

                            // Hide and reload list if success
                            dialog.dismiss();
                            if (response.getStatus().isSuccess())
                                activity.loadList();
                        }, request );
                })).create().show()
            );
        }

    }

}

File: BI-KOT/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/food/FoodAdapter.kt
package cz.wrzecond.restsys.food

import android.content.Intent
import android.graphics.Color
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import com.google.android.material.floatingactionbutton.FloatingActionButton
import cz.wrzecond.restsys.Adapter
import cz.wrzecond.restsys.R
import cz.wrzecond.restsys.allergen.AllergenActivity
import cz.wrzecond.restsys.databinding.ItemFoodBinding
import cz.wrzecond.restsys.task.HttpMethod
import cz.wrzecond.restsys.task.HttpRequest
import wrzecond.TjvFoodReadDTO
import java.util.*

class FoodAdapter (activity: FoodActivity, dtos: List<TjvFoodReadDTO>? = null,
val selecting: Boolean = false, errorResId: Int? = null) : Adapter<TjvFoodReadDTO>(activity, dtos, errorResId) {

    override fun createViewHolder (parent: ViewGroup)
        = ViewHolder(ItemFoodBinding.inflate(activity.layoutInflater, parent, false))

    inner class ViewHolder (private val binding: ItemFoodBinding) : Adapter<TjvFoodReadDTO>.ViewHolder(binding.root) {
        override fun update (dto: TjvFoodReadDTO) {
            // Basic information
            binding.foodName.text = dto.name
            binding.foodPrice.text = String.format(Locale.getDefault(), "%d,-", dto.price)

            // Allergen list
            if (dto.allergens.isNotEmpty()) {
                val sb = StringBuilder(activity.getString(R.string.food_allergens) + ":")
                dto.allergens.forEach { sb.append(" $it") }
                binding.allergenText.text = sb.toString()
            }
            else {
                binding.allergenText.visibility = View.GONE
                binding.allergenInfo.visibility = View.GONE
            }

            // Allergen info
            binding.allergenInfo.setOnClickListener {
                val intent = Intent(activity, AllergenActivity::class.java)
                        .putExtra(AllergenActivity.EXTRA_FOOD_ID, dto.id)
                activity.startActivity(intent)
            }

            // Select food on click
            if ( selecting ) {
                initSelection(dto)
                return
            }

            // Basic table view
            if ( !activity.service.isAdmin || !activity.intent.hasExtra(FoodActivity.EXTRA_SHOW_NAVIGATION) ) {
                binding.foodEdit.visibility = View.GONE
                binding.foodDelete.visibility = View.INVISIBLE
                return
            }

            // Admin view (edit + delete enabled)
            binding.foodEdit.setOnClickListener {
                AlertDialog.Builder(activity)
                    .setView(R.layout.create_food)
                    .create()
                    .apply { setOnShowListener(activity.createListener(this, dto)) }
                    .show()
            }
            binding.foodDelete.setOnClickListener {
                AlertDialog.Builder(activity)
                    .setTitle(R.string.food_delete)
                    .setMessage(activity.getString(R.string.food_delete_confirm, dto.name))
                    .setNegativeButton(R.string.button_dismiss, null)
                    .setPositiveButton(R.string.button_delete) { di,_ ->
                        val request = HttpRequest("/food/${dto.id}", HttpMethod.DELETE)
                        activity.service.scheduleTask(request) { response ->
                            Toast.makeText(activity, response.getToastString(activity), Toast.LENGTH_SHORT).show()
                            di.dismiss()
                            if (response.status.isSuccess)
                                activity.loadList()
                        }
                    }
                    .create()
                    .show()
            }
        }
        private fun initSelection (dto: TjvFoodReadDTO) {
            val activity = activity as FoodActivity

            // Add one
            binding.root.setOnClickListener {
                // Add one
                val originalCount = activity.selectedFood.getOrDefault(dto.id, 0)
                activity.selectedFood[dto.id] = originalCount + 1

                binding.foodCount.visibility = View.VISIBLE
                binding.foodCount.text = String.format(Locale.getDefault(), "%dx", originalCount + 1)

                binding.foodDelete.visibility = View.VISIBLE
                it.setBackgroundColor(Color.argb(127, 0, 255, 0))
                activity.findViewById<FloatingActionButton>(R.id.fab).visibility = View.VISIBLE
            }

            // Remove one
            binding.foodDelete.setOnClickListener {
                // Cannot decrement if null
                val originalCount = activity.selectedFood[dto.id] ?: return@setOnClickListener

                // Remove
                if (originalCount == 1) {
                    activity.selectedFood.remove(dto.id)
                    binding.root.setBackgroundColor(Color.rgb(255, 255, 255))
                    binding.foodCount.visibility = View.GONE
                    binding.foodDelete.visibility = View.INVISIBLE
                }
                // Just decrement
                else {
                    activity.selectedFood[dto.id] = originalCount - 1
                    binding.foodCount.text = String.format(Locale.getDefault(), "%dx", originalCount - 1)
                }

                activity.findViewById<FloatingActionButton>(R.id.fab).visibility = if (activity.selectedFood.isEmpty()) View.GONE else View.VISIBLE
            }

            val selectedCount = activity.selectedFood[dto.id]

            // Already selected
            if (selectedCount != null) {
                binding.foodCount.visibility = View.VISIBLE
                binding.foodCount.text = String.format(Locale.getDefault(), "%dx", selectedCount)
                binding.root.setBackgroundColor(Color.argb(127, 0, 255, 0))
            }

            binding.foodEdit.visibility = View.GONE
            binding.foodDelete.visibility = if (selectedCount != null) View.VISIBLE else View.INVISIBLE
            activity.findViewById<FloatingActionButton>(R.id.fab).visibility = if (activity.selectedFood.isEmpty()) View.GONE else View.VISIBLE
        }
    }

}